File: actuator_sim/service/service.py
from abc import ABC, abstractmethod
from actuator_sim.utils.config import ActuatorConfig
from actuator_sim import logger
from actuator_sim.models import (
    SetPositionRequest,
    SetVelocityRequest,
    SetTorqueRequest,
    CommandResponse,
    PositionResponse,
    VelocityResponse,
    CurrentResponse,
)


class Service(ABC):
    @abstractmethod
    async def start(self):
        pass

    @abstractmethod
    async def stop(self):
        pass

    # ── Command methods ──────────────────────────────────────────────────────

    @abstractmethod
    async def set_position(self, request: SetPositionRequest) -> CommandResponse:
        """Set the target position."""

    @abstractmethod
    async def set_velocity(self, request: SetVelocityRequest) -> CommandResponse:
        """Set the target velocity."""

    @abstractmethod
    async def set_torque(self, request: SetTorqueRequest) -> CommandResponse:
        """Set the target torque."""

    # ── Telemetry methods ────────────────────────────────────────────────────

    @abstractmethod
    async def read_position(self) -> PositionResponse:
        """Return the current position."""

    @abstractmethod
    async def read_velocity(self) -> VelocityResponse:
        """Return the current velocity."""

    @abstractmethod
    async def read_current(self) -> CurrentResponse:
        """Return the current draw."""


class AppService(Service):
    def __init__(self, config: ActuatorConfig):
        self.config = config
        self._position: float = 0.0
        self._velocity: float = 0.0
        self._torque: float = 0.0

    async def start(self):
        logger.info("Starting App Service")

    async def stop(self):
        logger.info("Stopping App Service")

    async def set_position(self, request: SetPositionRequest) -> CommandResponse:
        logger.debug("set_position: {}", request.angle)
        self._position = request.angle
        return CommandResponse(success=True, message="Position set")

    async def set_velocity(self, request: SetVelocityRequest) -> CommandResponse:
        logger.debug("set_velocity: {}", request.velocity)
        self._velocity = request.velocity
        return CommandResponse(success=True, message="Velocity set")

    async def set_torque(self, request: SetTorqueRequest) -> CommandResponse:
        logger.debug("set_torque: {}", request.torque)
        self._torque = request.torque
        return CommandResponse(success=True, message="Torque set")

    async def read_position(self) -> PositionResponse:
        return PositionResponse(angle=self._position)

    async def read_velocity(self) -> VelocityResponse:
        return VelocityResponse(velocity=self._velocity)

    async def read_current(self) -> CurrentResponse:
        # Stub: real hardware would query a current sensor
        return CurrentResponse(current=0.0)

File: smart-actuator/crates/actuator-core/src/service.rs
use tracing::{debug, info};

use crate::types::{CommandResponse, CurrentResponse, PositionResponse, VelocityResponse};

/// The actuator service contract — transport-independent.
/// Both the real firmware and the simulator implement this trait.
#[async_trait::async_trait]
pub trait Service: Send + Sync {
    async fn start(&self);
    async fn stop(&self);

    // ── Level 1: command RPCs ────────────────────────────────────────────────

    /// Set the target position (radians).
    async fn set_position(&self, angle: f64) -> CommandResponse;

    /// Set the target velocity (rad/s).
    async fn set_velocity(&self, velocity: f64) -> CommandResponse;

    /// Set the target torque (N·m).
    async fn set_torque(&self, torque: f64) -> CommandResponse;

    // ── Level 1: telemetry RPCs ──────────────────────────────────────────────

    /// Return the current position.
    async fn read_position(&self) -> PositionResponse;

    /// Return the current velocity.
    async fn read_velocity(&self) -> VelocityResponse;

    /// Return the current draw.
    async fn read_current(&self) -> CurrentResponse;
}

/// Stub implementation — stores the last commanded value, no dynamics.
/// This is the current-generation simulator. A dynamics model will replace
/// the state fields here once physics is added (RFD-3 S2).
pub struct AppService {
    // Interior mutability via a lightweight async-aware mutex so the service
    // can be shared across tonic's multi-threaded executor without cloning.
    state: tokio::sync::Mutex<ActuatorState>,
}

struct ActuatorState {
    position: f64,
    velocity: f64,
    torque: f64,
}

impl AppService {
    pub fn new() -> Self {
        Self {
            state: tokio::sync::Mutex::new(ActuatorState {
                position: 0.0,
                velocity: 0.0,
                torque: 0.0,
            }),
        }
    }
}

impl Default for AppService {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait::async_trait]
impl Service for AppService {
    async fn start(&self) {
        info!("AppService started");
    }

    async fn stop(&self) {
        info!("AppService stopped");
    }

    async fn set_position(&self, angle: f64) -> CommandResponse {
        debug!(angle, "set_position");
        self.state.lock().await.position = angle;
        CommandResponse {
            success: true,
            message: "Position set".into(),
        }
    }

    async fn set_velocity(&self, velocity: f64) -> CommandResponse {
        debug!(velocity, "set_velocity");
        self.state.lock().await.velocity = velocity;
        CommandResponse {
            success: true,
            message: "Velocity set".into(),
        }
    }

    async fn set_torque(&self, torque: f64) -> CommandResponse {
        debug!(torque, "set_torque");
        self.state.lock().await.torque = torque;
        CommandResponse {
            success: true,
            message: "Torque set".into(),
        }
    }

    async fn read_position(&self) -> PositionResponse {
        PositionResponse {
            angle: self.state.lock().await.position,
        }
    }

    async fn read_velocity(&self) -> VelocityResponse {
        VelocityResponse {
            velocity: self.state.lock().await.velocity,
        }
    }

    async fn read_current(&self) -> CurrentResponse {
        // Stub: no current sensor in simulation yet.
        CurrentResponse { current: 0.0 }
    }
}

// ── Unit tests ───────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    fn svc() -> AppService {
        AppService::new()
    }

    #[tokio::test]
    async fn set_and_read_position() {
        let s = svc();
        let resp = s.set_position(1.23).await;
        assert!(resp.success);
        assert_eq!(s.read_position().await.angle, 1.23);
    }

    #[tokio::test]
    async fn set_and_read_velocity() {
        let s = svc();
        let resp = s.set_velocity(0.5).await;
        assert!(resp.success);
        assert_eq!(s.read_velocity().await.velocity, 0.5);
    }

    #[tokio::test]
    async fn set_and_read_torque() {
        let s = svc();
        let resp = s.set_torque(2.0).await;
        assert!(resp.success);
        // torque is stored; no dedicated read_torque yet — verify via state
        assert_eq!(resp.message, "Torque set");
    }

    #[tokio::test]
    async fn read_current_is_zero() {
        let s = svc();
        assert_eq!(s.read_current().await.current, 0.0);
    }

    #[tokio::test]
    async fn initial_position_is_zero() {
        let s = svc();
        assert_eq!(s.read_position().await.angle, 0.0);
    }
}
